exporter/metric/metric.go

```diff
@@ -0,0 +1,384 @@
+	errUnsupportedAggregation = errors.New("Currently the metric type is not supported")
+)
+
+// metricExporter is the implementation of OpenTelemetry metric exporter for
+// Google Cloud Monitoring.
+type metricExporter struct {
+	o *options
+
+	// mdCache is the cache to hold MetricDescriptor to avoid creating duplicate MD.
+	// TODO: [ymotongpoo] this map should be goroutine safe with mutex.
+	mdCache map[string]*googlemetricpb.MetricDescriptor
+
+	client *monitoring.MetricClient
+}
+
+// newMetricExporter returns an exporter that uploads OTel metric data to Google Cloud Monitoring.
+// Only one Cloud Monitoring exporter should be created per ProjectID per process, any subsequent
+// invocations of NewExporter with the same ProjectID will return an error.
+func newMetricExporter(o *options) (*metricExporter, error) {
+	if strings.TrimSpace(o.ProjectID) == "" {
+		return nil, errBlankProjectID
+	}
+
+	clientOpts := append(o.MonitoringClientOptions, option.WithUserAgent(userAgent))
+	ctx := o.Context
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	client, err := monitoring.NewMetricClient(ctx, clientOpts...)
+	if err != nil {
+		return nil, err
+	}
+	e := &metricExporter{
+		client: client,
+		o:      o,
+	}
+	return e, nil
+}
+
+// InstallNewPipeline instantiates a NewExportPipeline and registers it globally.
+func InstallNewPipeline(popts ...push.Option) (*push.Controller, error) {
+	pusher, err := NewExportPipeline(popts...)
+	if err != nil {
+		return nil, err
+	}
+	global.SetMeterProvider(pusher)
+	return pusher, err
+}
+
+// NewExportPipeline sets up a complete export pipeline with the recommended setup,
+// chaining a NewRawExporter into the recommended selectors and integrators.
+func NewExportPipeline(popts ...push.Option) (*push.Controller, error) {
+	selector := simple.NewWithExactMeasure()
+	exporter, err := NewRawExporter()
+	if err != nil {
+		return nil, err
+	}
+	integrator := integrator.New(selector, true)
+	period := exporter.metricExporter.o.ReportingInterval
+	pusher := push.New(integrator, exporter, period, popts...)
+	return pusher, nil
+}
+
+// ExportMetrics exports OpenTelemetry Metrics to Google Cloud Monitoring.
+func (me *metricExporter) ExportMetrics(ctx context.Context, res *resource.Resource, cps export.CheckpointSet) error {
+	if err := me.exportMetricDescriptor(ctx, cps); err != nil {
+		return err
+	}
+
+	if err := me.exportTimeSeries(ctx, res, cps); err != nil {
+		return err
+	}
+	return nil
+}
+
+// exportMetricDescriptor create MetricDescriptor from the record
+// if the descriptor is not registerd in Cloud Monitoring yet.
+func (me *metricExporter) exportMetricDescriptor(ctx context.Context, cps export.CheckpointSet) error {
+	mds := make(map[string]*googlemetricpb.MetricDescriptor)
+	aggError := cps.ForEach(func(r export.Record) error {
+		desc := r.Descriptor()
+		name := desc.Name()
+
+		if _, ok := me.mdCache[name]; !ok {
+			if _, localok := mds[name]; !localok {
```

Comment:
The nesting can be reduced here via reverse conditions, i.e.
```
if _, ok := map[key]; ok {
  // skip, already present
  return
}
...
```

Reply:
whoa! yes, I reversed the condition and extracted nested `if` from `mdCache` checking block.